pkg/expr/convert_to_long.go

```diff
@@ -0,0 +1,309 @@
+package expr
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+func ConvertToLong(frames data.Frames) (data.Frames, error) {
+	if len(frames) == 0 {
+		// general empty case for now
+		return frames, nil
+	}
+	// Four Conversion Possible Cases
+	// 1. NumericMulti -> NumericLong
+	// 2. NumericWide -> NumericLong
+	// 3. TimeSeriesMulti -> TimeSeriesLong
+	// 4. TimeSeriesWide -> TimeSeriesLong
+
+	// Detect if input type is declared
+	// First Check Frame Meta Type
+
+	var inputType data.FrameType
+	if frames[0].Meta != nil && frames[0].Meta.Type != "" {
+		inputType = frames[0].Meta.Type
+	}
+	if inputType != "" && !supportedToLongConversion(inputType) {
+		return frames, fmt.Errorf("unsupported input type %s for SQL expression", inputType)
+	}
+
+	// TODO: Add some guessing of Type if not declared
+
+	if inputType == "" {
+		return frames, fmt.Errorf("could not determine input type")
+	}
+
+	convert := getToLongConversionFunc(inputType)
+	if convert == nil {
+		return frames, fmt.Errorf("could not get conversion function for input type %s", inputType)
+	}
+
+	return convert(frames)
+}
+
+func convertNumericMultiToNumericLong(frames data.Frames) (data.Frames, error) {
+	// Apart from metadata. NumericMulti is basically numeric Wide, except one frame per thing
+	// to we collapse into one frame and call the wide conversion
+	newFrame := data.NewFrame("")
+	for _, frame := range frames {
+		for _, field := range frame.Fields {
+			if !field.Type().Numeric() {
+				continue
```

Comment:
This feels dangerous to me - ~~like we'll throw away data, and we won't know we're doing it.~~
I wonder if we can return an error in this case, or log an error, or something.

Isn't this an error case in fact - aren't numeric-multi supposed to only have a single field?
(I checked your DataPlane docs [here](https://grafana.com/developers/dataplane/numeric#numeric-multi-format-numericmulti) and I can see you have made a provision for "Remainder data". I'm wondering if we really need this -I feel like we'd do well to tighten up the spec and eliminate this "Remainder data" entirely, but I expect that's outside the scope of this PR)

Reply:
We can attach a warnings to the frame returned